### PR TITLE
Add validation to widget.Form

### PR DESCRIPTION
**Pull request by Jacalz**
### Description:
This adds necessary changes for making sure that form handles validation in widget.Form.

Partial fix to #916 

### Checklist:

- [ ] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:

- [x] Public APIs match existing style.
- [x] Any breaking changes have a deprecation path or have been discussed.

### CODE REVIEW

File: widget/form.go [hunk 1]
```diff
@@ -107,6 +110,30 @@ func (f *Form) updateButtons() {
...
+		}
+	}
+}
+
+func (f *Form) setUpValidation(widget fyne.CanvasObject, i int) {
+	var w interface{} = widget
+	if w, ok := w.(fyne.Validatable); ok {
+		w.SetOnValidationChanged(func(err error) {
+			f.Items[i].validationError = err
+			if f.Items[i].validationError != nil {
```
Comment (stuartmscott) [inline widget/form.go:130]:
```suggestion
			if err != nil {
```

File: cmd/fyne_demo/screens/widget.go [hunk 2]
```diff
@@ -225,10 +225,14 @@ func makeProgressTab() fyne.CanvasObject {
 func makeFormTab() fyne.Widget {
 	name := widget.NewEntry()
 	name.SetPlaceHolder("John Smith")
+
 	email := widget.NewEntry()
 	email.SetPlaceHolder("test@example.com")
+	email.Validator = validation.NewRegexp(`\w{1,}@\w{1,}\.\w{1,4}`, "not a valid email")
```
[the remark this answers is not in the document]
Comment (Jacalz) [reply cmd/fyne_demo/screens/widget.go:231]: I think that would be something for another PR. That regexp is overly simplified.

File: widget/entry_validation.go [hunk 3]
```diff
@@ -7,6 +7,25 @@ import (
...
+func (e *Entry) Validate() error {
+	if e.Validator == nil {
+		return nil
+	}
+
+	return e.Validator(e.Text)
+}
+
+// SetOnValidationChanged is intended for parent widgets or containers to hook into the validation.
+// Should not be used by a developer specifically.
```
[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/entry_validation.go:22]: Yeah perhaps. I just felt that a developer might wonder why their own callback doesn't work if the widget is in a form for example.

[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/entry_validation.go:22]: Perhaps we just need to specify that the function might be overwritten by a parent widget?

File: cmd/fyne_demo/screens/widget.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartmscott) [reply cmd/fyne_demo/screens/widget.go:231]: Ack

File: widget/entry_validation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (stuartmscott) [reply widget/entry_validation.go:22]: Or if there is an existing callback, then create a new function that calls both and make that the new callback

[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/entry_validation.go:22]: That is a good idea. The only problem is that the widget might be saving the function anywhere. I guess we could do that in entry, but that doesn't sound like a perfect solution. What do you think?

File: widget/form.go [hunk 4]
```diff
@@ -107,6 +110,30 @@ func (f *Form) updateButtons() {
...
+	f.checkValidation()
+}
+
+func (f *Form) checkValidation() {
+	for i, item := range f.Items {
+		if item.validationError == nil && i == len(f.Items)-1 {
+			f.submitButton.Enable()
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/form.go:123]: Good catch. Looks like I had accidentally removed a piece of code form there.

File: widget/entry.go [hunk 5]
```diff
@@ -1130,17 +1131,28 @@ func (r *entryRenderer) Refresh() {
...
+
 	if r.entry.Validator != nil {
-		r.entry.validationError = r.entry.Validator(content)
+		err := r.entry.Validator(content)
+		if err != r.entry.validationError {
+			r.entry.validationError = err
+
+			if r.entry.onValidationChanged != nil {
+				r.entry.onValidationChanged(r.entry.validationError)
+			}
```
[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/entry.go:1142]: Good idea.

File: widget/form.go [hunk 6]
```diff
@@ -107,6 +110,30 @@ func (f *Form) updateButtons() {
...
+
+func (f *Form) setUpValidation(widget fyne.CanvasObject, i int) {
+	var w interface{} = widget
+	if w, ok := w.(fyne.Validatable); ok {
```
[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/form.go:127]: Good  catch. I must have been doing something strange yesterday.

File: widget/form.go [hunk 7]
```diff
@@ -10,13 +10,14 @@ import (
...
-	Widget fyne.CanvasObject
+	Text            string
+	Widget          fyne.CanvasObject
+	validationError error
 }
 
 // NewFormItem creates a new form item with the specified label text and input widget
 func NewFormItem(text string, widget fyne.CanvasObject) *FormItem {
-	return &FormItem{text, widget}
+	return &FormItem{text, widget, nil}
```
[the remark this answers is not in the document]
Comment (stuartmscott) [reply widget/form.go]: Sorry I wasn't clear enough. By switching to named fields, we no longer need to initialize to nil as that is the default

File: widget/entry_validation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (stuartmscott) [reply widget/entry_validation.go]: I think your latest change to the comment makes it clear

File: widget/form.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jacalz) [reply widget/form.go]: Oh right. Forgot about that :I

File: widget/entry.go [hunk 8]
```diff
@@ -1130,17 +1131,28 @@ func (r *entryRenderer) Refresh() {
...
 	}
+
 	if r.entry.Validator != nil {
-		r.entry.validationError = r.entry.Validator(content)
+		err := r.entry.Validator(content)
+		if err != r.entry.validationError {
+			r.entry.validationError = err
+
+			if f := r.entry.onValidationChanged; f != nil {
+				f(r.entry.validationError)
```
Comment (stuartmscott) [inline widget/entry.go:1141]: nit:
```suggestion
				f(err)
```

**Review by stuartmscott — APPROVED** [review]

**Review by andydotxyz — DISMISSED** [review]
Very neat solution thanks :).
I agree EmailValidator would be good to add :)

**Review by stuartmscott — APPROVED** [review]

**Pull request merged**